python/ray/experimental/dag/dag_node.py

```diff
@@ -0,0 +1,217 @@
+    """Abstract class for a node in a Ray task graph.
+
+    A node has a type (e.g., TaskNode), data (e.g., function options and body),
+    arguments (Python values, DAGNodes, and DAGNodes nested within Python
+    argument values) and options (Ray API .options() used for function, class
+    or class method)
+    """
+
+    def __init__(
+        self,
+        args: Tuple[Any],
+        kwargs: Dict[str, Any],
+        options: Dict[str, Any],
+        kwargs_to_resolve: Optional[Dict[str, Any]] = None,
+    ):
+        # Bound node arguments, ex: func_or_class.bind(1)
+        self._bound_args: Tuple[Any] = [] if args is None else args
+        # Bound node keyword arguments, ex: func_or_class.bind(a=1)
+        self._bound_kwargs: Dict[str, Any] = {} if kwargs is None else kwargs
+        # Bound node options arguments, ex: func_or_class.options(num_cpus=2)
+        self._bound_options: Dict[str, Any] = {} if options is None else options
+        # Bound kwargs to resolve that's specific to subclass implementation
+        self._bound_kwargs_to_resolve: Optional[Dict[str, Any]] = kwargs_to_resolve
+        # UUID that is not changed over copies of this node.
+        self._stable_uuid = uuid.uuid4().hex
+
+    def get_args(self) -> Tuple[Any]:
+        """Return the tuple of arguments for this node."""
+
+        return self._bound_args
+
+    def get_kwargs(self) -> Dict[str, Any]:
+        """Return the dict of keyword arguments for this node."""
+
+        return self._bound_kwargs.copy()
+
+    def get_options(self) -> Optional[Dict[str, Any]]:
+        """Return the dict of options arguments for this node."""
+
+        return self._bound_options.copy()
+
+    def execute(self) -> Union[ray.ObjectRef, ray.actor.ActorHandle]:
+        """Execute this DAG using the Ray default executor."""
+        return self._apply_recursive(lambda node: node._execute_impl())
+
+    def _get_toplevel_child_nodes(self) -> Set["DAGNode"]:
+        """Return the set of nodes specified as top-level args.
+
+        For example, in `f.remote(a, [b])`, only `a` is a top-level arg.
+
+        This set of nodes are those that are typically resolved prior to
+        task execution in Ray. This does not include nodes nested within args.
+        For that, use ``_get_all_child_nodes()``.
+        """
+
+        children = set()
+        for a in self.get_args():
+            if isinstance(a, DAGNode):
+                children.add(a)
+        for a in self.get_kwargs().values():
+            if isinstance(a, DAGNode):
+                children.add(a)
+        return children
+
+    def _get_all_child_nodes(self) -> Set["DAGNode"]:
+        """Return the set of nodes referenced by the args of this node.
+
+        For example, in `f.remote(a, [b])`, this includes both `a` and `b`.
+        """
+
+        f = _PyObjScanner()
+        children = set()
+        for n in f.find_nodes(
+            [self._bound_args, self._bound_kwargs, self._bound_options]
```

Comment:
Is it valid to pass a node into `bound_options`? If there's no requirement for it, I'd suggest we don't support this for now to avoid added complexity

Reply:
Looks like it's untested as well, let's either explicitly support & test it or not.

Reply:
for bound_options, no, it's just runtime configurations and we can remove it.